Add cp2k tutorial (#1935)
inductiva/inductiva: tutorials/simulators/CP2K.md

This guide will walk you through setting up and running CP2K simulations using the Inductiva API.

> # CP2K

Reviewer: We should mention the AVX part

[CP2K](https://www.cp2k.org/) is an open-source quantum chemistry and
solid-state physics software designed for atomistic simulations. It supports a
wide range of methods, including density functional theory (DFT), semi-empirical
approaches, and classical force fields, making it suitable for molecular
dynamics, electronic structure calculations, and materials science applications.
Its flexibility and extensive features make it a powerful tool for researchers
in computational chemistry, condensed matter physics, and related fields.

## Supported Versions
We currently support the following CP2K version:
- **v2025.1**

## Running the CP2K H2O-64 Benchmark

### Objective

This tutorial will demonstrate how to run a CP2K simulation using the H2O-64
benchmark, which simulates a system that consists of 64 water molecules in a
12.4 Å³ cell, with MD running for 1000 steps.

### Prerequisites  

To follow this tutorial, download the input file for the H2O-64 benchmark from
[here](https://github.com/cp2k/cp2k/blob/master/benchmarks/QS/H2O-64.inp) and
place it in a folder called `H2O-64`. Once you have the simulation file, you're

> ready to scale your simulations to the cloud.

Reviewer: ```suggestion
ready to scale your simulations to the Cloud.
```

### Increasing the simulation time

> The H2O-64 benchmark, as configured, runs for only 10 steps, which corresponds
> to a simulated time of 5 femtoseconds. To obtain more relevant results, we
> increase the simulation time by 100x, extending it to 500 femtoseconds. This
> requires increasing the number of MD steps from 10 to 1000 in the CP2K input file.
> 
> To do that open the `H2O-64.inp` file and change `STEPS 10` to `STEPS 1000`.

Reviewer: ```suggestion
The H2O-64 benchmark only runs for 10 steps as configured, which corresponds to a simulated time of 5 femtoseconds. For the purposes of this tutorial, we decided to increase the simulation time by a factor of 100 to 500 femtoseconds. This requires increasing the number of MD steps from 10 to 1000 in the CP2K input file.

To do this, open the file `H2O-64.inp` and change `STEPS 10` to `STEPS 1000`.
```

## Supported Versions
We currently support the following CP2K version:
- **v2025.1** - Compiled with AVX2.

## Running the CP2K H2O-64 Benchmark

### Objective

This tutorial will show you how to run a CP2K simulation using the H20-64
benchmark available on the [official CP2K website](https://www.cp2k.org/performance#benchmarks).
This use case simulates a system containing 64 water molecules (192 atoms,
512 electrons) in a 12.4 Å³ cell, with MD running for 10 steps.

We will also demonstrate Inductiva's ability to efficiently scale this use case
on a more powerful machine.

### Prerequisites  

To follow this tutorial, download the input file for the H2O-64 benchmark from
[here](https://github.com/cp2k/cp2k/blob/master/benchmarks/QS/H2O-64.inp) and
place it in a folder called `H2O-64`. Once you have the simulation file, you're
ready to scale your simulations to the Cloud.

### Running Your Simulation

Here is the code required to run a CP2K simulation using the Inductiva API.

We will be running this simulation on a 16 vCPU virtual machine supported by a
4th generation AMD EPYC™ (Genoa) processor.

```python
"""CP2K Simulation."""
import inductiva

> # Instantiate machine group

Reviewer 2: Instantiate a machine on Google Cloud Platform

Reviewer: “Allocate” is the word we’ve been using in this comment

> We will be running this simulation on a 16 vCPU virtual machine supported by a
> 4th generation AMD EPYC™ (Genoa) processor.

Reviewer: ```suggestion
In this example, we will execute the simulation on a virtual machine with 16 vCPUs, supported by a 4th generation AMD EPYC™ (Genoa) processor.
```

cloud_machine = inductiva.resources.MachineGroup( 
    provider="GCP",
    machine_type="c3d-highcpu-16")

# Initialize the Simulator
cp2k = inductiva.simulators.CP2K( 
    version="2025.1")

# Run simulation
task = cp2k.run( 
    input_dir="/Path/to/H2O-64",
    sim_config_filename="H2O-64.inp",

>     n_vcpus=16,

Reviewer 2: Do we need this?
I really think there is no need for this parameter, especially if we are running everything with the same number of vCPUs. If we were benchmarking options, and trying different n_vcpus for the same machine, yes, it would make sense to show this. But let's try to avoid bringing complexity for the user. The defaults should be reasoable.

Author: I added this due to this issue https://github.com/inductiva/tasks/issues/486

Should we drop this issue and close it ?

>     use_hwthread=True,

Reviewer 2: Do we need this?
Same comment as before.
(and please propagate to other tutorials that do not require these flags).
Once we have the API docs ready (sphinx generated) it may be worth telling the user about this parameter.

Timeline:
	Waiting for Input         at 05/03, 16:04:17      1.525 s
	In Queue                  at 05/03, 16:04:19      21.308 s
	Preparing to Compute      at 05/03, 16:04:40      8.307 s
	In Progress               at 05/03, 16:04:48      103.301 s
		└> 103.177 s       /opt/openmpi/5.0.6/bin/mpirun --use-hwthread-cpus --np 16 cp2k.psmp H2O-64.inp
	Finalizing                at 05/03, 16:06:32      0.453 s
	Success                   at 05/03, 16:06:32      

Data:
	Size of zipped output:    87.62 KB
	Size of unzipped output:  296.14 KB
	Number of output files:   6

Estimated computation cost (US$): 0.0049 US$

Go to https://console.inductiva.ai/tasks/7y33ri1umg0tsppabxf5jska4 for more details.
```

The core computation time for this simulation was approximately **1 minute and 43 seconds**
(103 seconds), as shown in the `In Progress` line. This represents the
actual execution time of the CP2K benchmark on a 16 virtual CPU machine.

For comparison, the same simulation takes **1 minute and 15 seconds** on a similar
local machine with a 16 virtual CPUs (Ryzen 7 7700X). This performance
difference is expected, as cloud CPUs typically have lower clock speeds compared to

> high-performance desktop processors, prioritizing energy efficiency and density

Reviewer 2: compared to regular desktop processors.

This is not a high-performance CPU :)

over raw speed.

> However, increasing the number of vCPUs on this cloud machine can improve this
> result.

Reviewer: ```suggestion
However, increasing the number of vCPUs on the cloud machine could enhance performance.
```

two parameters:

1. Modify the `machine_type` to a more powerful machine with more vCPUs.
2. Adjust the `n_vcpus` accordingly to maximize parallel processing efficiency.

We tested this simulation across multiple machines to analyze how performance
and cost scale with increasing computational resources.

As mentioned earlier, running this simulation on a **16 vCPU** cloud machine was
slower than on a similarly powered local computer. To enhance performance, we
upgraded to a **c3d-highcpu-60** instance with **60 vCPUs**, reducing the
runtime to just **38 seconds**—with a slight cost increase to **$0.0077**.

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|-------------------------|--------------|------------------|---------------|
| **Local Ryzen 7 7700X** | 16           | 1 minute and 15 seconds | N/A           |
| **Cloud c3d-highcpu-16** | 16           | 1 minute and 43 seconds | 0.0049 US$      |
| **Cloud c3d-highcpu-60** | 60           | 38 seconds | 0.0077 US$      | 

By leveraging the Inductiva API, you can efficiently scale your CP2K simulations
to meet your computational needs. Try different machine configurations and

> optimize your workflow for faster and more cost-effective results!

Reviewer: ```suggestion
optimize your workflow for faster, more cost-effective results!
```

over raw speed.

However, increasing the number of vCPUs on this cloud machine can improve this
result.

### Scaling Up Your Simulation

With the Inductiva API, scaling up your CP2K simulation is as simple as changing
two parameters:

1. Modify the `machine_type` to a more powerful machine with more vCPUs.
2. Adjust the `n_vcpus` accordingly to maximize parallel processing efficiency.

We tested this simulation across multiple machines to analyze how performance
and cost scale with increasing computational resources.

As mentioned earlier, running this simulation on a **16 vCPU** cloud machine was
slower than on a similarly powered local computer. To enhance performance, we
upgraded to a **c3d-highcpu-60** instance with **60 vCPUs**, reducing the
runtime to just **38 seconds**—with a slight cost increase to **$0.0077**.

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|------------------------|--------------|------------------|---------------|
| **Cloud c3d-highcpu-16** | 16           | 1 minute and 43 seconds | 0.0049 US$      |
| **Cloud c3d-highcpu-60** | 60           | 38 seconds | 0.0077 US$      |

> 

Reviewer: We can remove this, so it isn't repetitive

### **Final Comparison**  

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|-------------------------|--------------|------------------|---------------|
| **Local Ryzen 7 7700X** | 16           | 1 minute and 15 seconds | N/A           |
| **Cloud c3d-highcpu-16** | 16           | 1 minute and 43 seconds | 0.0049 US$      |
| **Cloud c3d-highcpu-60** | 60           | 38 seconds | 0.0077 US$      | 

By leveraging the Inductiva API, you can efficiently scale your CP2K simulations
to meet your computational needs. Try different machine configurations and

> optimize your workflow for faster and more cost-effective results!

Reviewer: and have only this table in the previous subsection

> This guide will walk you through setting up and running CP2K simulations using the Inductiva API.

Reviewer 2: This guide will walk you through running your first CP2K simulations using the Inductiva API.

# CP2K

[CP2K](https://www.cp2k.org/) is an open-source quantum chemistry and
solid-state physics software designed for atomistic simulations. It supports a

> wide range of methods, including density functional theory (DFT), semi-empirical

Reviewer 2: Density Functional Theory (caps)

> approaches, and classical force fields, making it suitable for molecular

Reviewer 2: remove , before "and"

> approaches, and classical force fields, making it suitable for molecular

Reviewer 2: I think we should remove "approaches"

dynamics, electronic structure calculations, and materials science applications.
Its flexibility and extensive features make it a powerful tool for researchers
in computational chemistry, condensed matter physics, and related fields.

## Supported Versions
We currently support the following CP2K version:

> - **v2025.1** - Compiled with AVX2.

Reviewer 2: Does it run on GPU?

Author: Yes, with the current limitations known to the simulator. Should i add `Compiled with cuda and AVX2` ?

## Running the CP2K H2O-64 Benchmark

### Objective

> This tutorial will demonstrate how to run a CP2K simulation using the H2O-64

Reviewer 2: This tutorial will demonstrate how to run the H2O-64 benchmark....

benchmark, which simulates a system that consists of 64 water molecules in a

> 12.4 Å³ cell, with MD running for 10 steps.

Reviewer 2: Yes: 64 water molecules is really really really small. 
These simulations are supposed to have hundreds of thousands of molecules.
I am pretty sure that this would NOT make any interesting use of a GPU....

### Prerequisites  

To follow this tutorial, download the input file for the H2O-64 benchmark from
[here](https://github.com/cp2k/cp2k/blob/master/benchmarks/QS/H2O-64.inp) and

> place it in a folder called `H2O-64`. Once you have the simulation file, you're

Reviewer 2: which simulation file?

Author: The one i pointed in the link. The `here` link points directly to a file.

ready to scale your simulations to the Cloud.

### Running Your Simulation

Here is the code required to run a CP2K simulation using the Inductiva API:

> 

Reviewer 2: WE will be running this simulation on a 16 vCPU virtual machine supported by a AMD Epyc XYZ processor (the cd3 machine series).

```python
"""CP2K Simulation."""
import inductiva

# Instantiate machine group
cloud_machine = inductiva.resources.MachineGroup( 
    provider="GCP",

>     machine_type="c3d-standard-16")

Reviewer 2: Can you run it on highcpu? If so, it probably makes more sense because it is much cheaper.
Also make the parameter spot explicit.

# Initialize the Simulator
cp2k = inductiva.simulators.CP2K( 
    version="2025.1")

# Run simulation
task = cp2k.run( 
    input_dir="/Path/to/H2O-64",
    sim_config_filename="H2O-64.inp",
    n_vcpus=16,
    use_hwthread=True,
    on=cloud_machine)

task.wait()

> cloud_machine.terminate()

Reviewer 2: This terminate() call really needs to be removed.

Author: I already have a PR to fix this. Once it's merged will go looking for terminates on the docs/tutorials to remove.

Once the simulation is complete, we terminate the machine, download the results,
and print a summary of the simulation as shown below.

```
inductiva tasks info ngjrax2xsak3yaonirxj1pdcc

Task status: Success

Timeline:
	Waiting for Input         at 05/03, 10:28:59      1.193 s
	In Queue                  at 05/03, 10:29:00      12.325 s
	Preparing to Compute      at 05/03, 10:29:13      6.645 s
	In Progress               at 05/03, 10:29:19      103.302 s
		└> 103.17 s        /opt/openmpi/4.1.6/bin/mpirun --use-hwthread-cpus --np 16 cp2k.psmp H2O-64.inp
	Finalizing                at 05/03, 10:31:03      0.447 s
	Success                   at 05/03, 10:31:03      

Data:
	Size of zipped output:    87.65 KB
	Size of unzipped output:  296.30 KB
	Number of output files:   6

Estimated computation cost (US$): 0.0058 US$

Go to https://console.inductiva.ai/tasks/ngjrax2xsak3yaonirxj1pdcc for more details.
```

The core computation time for this simulation was approximately 1 minute and 43
seconds (103 seconds), as shown in the `In Progress` line. This represents the
actual execution time of the CP2K benchmark on a 16 virtual CPU machine.

> 

Reviewer 2: For comparison, 
... local machine .....

### Scaling Up Your Simulation  

Scaling up your CP2K simulation is as simple as changing two parameters:

1. Modify the `machine_type` to a more powerful machine with more vCPUs.
2. Adjust the `n_vcpus` accordingly to maximize parallel processing efficiency.

We tested this simulation across multiple machines to analyze how performance
and cost scale with increasing computational resources.  

> We began with a local run on a **Ryzen 7 7700X** with **16 vCPUs**, completing

Reviewer 2: This reference should be moved up before we talk about scaling.
The "story" would be that we know that the local machine would be a bit faster. And we should take the time to explain why. Then we can say, "The good thing about Inductiva is that you only need to change one line of code to speed up your simulation". And then you start this section.

the simulation in **1 minute and 15 seconds**. To compare this with a
cloud-based machine of similar specifications, we used a **c3d-standard-16**
machine, which also has **16 vCPUs**. However, the cloud machine was slower,
taking **1 minute and 50 seconds**, with a cost of **0.0058 US$**. This
performance difference is expected, as cloud CPUs typically have lower clock

> speeds compared to high-performance desktop processors, prioritizing energy

Reviewer 2: Move up....

efficiency and density over raw speed.

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|-------------------------|--------------|------------------|---------------|
| **Local Ryzen 7 7700X** | 16           | 1 minute and 15 seconds | N/A           |
| **Cloud c3d-standard-16** | 16           | 1 minute and 50 seconds | 0.0058 US$      |

To improve performance, we scaled up to a **c3d-standard-60** machine with
**60 vCPUs**. This significantly reduced the runtime to **51 seconds**, with the
cost increasing slightly to **0.0098 US$**.  

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|------------------------|--------------|------------------|---------------|
| **Cloud c3d-standard-16** | 16           | 1 minute and 50 seconds | 0.0058 US$      |

> | **Cloud c3d-standard-60** | 60           | 51 seconds | 0.0098 US$      |

Reviewer 2: Add c4 machines too please.

> To further explore cloud scaling, we tested two additional machines:

Reviewer 2: This is another section: How much is too much?

performance. The results show worst results, with runtimes increasing to
**1 minute and 5 seconds** and **1 minute and 48 seconds**, while costs surged
to **0.037 US$** and **0.12 US$**, respectively.  

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|------------------------|--------------|------------------|---------------|
| **Cloud c3d-standard-180** | 180           | 1 minute and 5 seconds | 0.037 US$      |
| **Cloud c3d-standard-360** | 360           | 1 minute and 48 seconds | 0.12 US$      |

### **Final Comparison**  

| Machine Type            | Virtual CPUs | Time              | Estimated Cost |
|-------------------------|--------------|------------------|---------------|
| **Local Ryzen 7 7700X** | 16           | 1 minute and 15 seconds | N/A           |
| **Cloud c3d-standard-16** | 16           | 1 minute and 50 seconds | 0.0058 US$      |
| **Cloud c3d-standard-60** | 60           | 51 seconds | 0.0098 US$      |
| **Cloud c3d-standard-180** | 180           | 1 minute and 5 seconds | 0.037 US$      |
| **Cloud c3d-standard-360** | 360           | 1 minute and 48 seconds | 0.12 US$      |

### **Key Takeaway**  

> Selecting the right machine for your simulation is essential for balancing

Reviewer 2: No... This is because the simulation is too small.....
This is not a good takeway.
If we are not comparing the performance of this simulation with some other values, please lets go to a much bigger simulation. This is not a good "selling" document.

Author: I see this tutorial as the validation tutorial and not a selling document. For a selling document we can run the benchmark [H2O-8192.inp](https://github.com/cp2k/cp2k/blob/master/benchmarks/QS/H2O-8192.inp) (maybe a smaller one).

This simulation might not even be possible to run on a local machine due to lack of memory.

> This guide will walk you through running your first CP2K simulations using the Inductiva API.

Author: Gx ?
